python/cuml/dask/preprocessing/encoders.py

```diff
@@ -0,0 +1,357 @@
+        # values, there will be ambiguous cells. This creates difficulties
+        # in interpreting the model.
+        if self.drop is not None and self.handle_unknown != 'error':
+            raise ValueError(
+                "`handle_unknown` must be 'error' when the drop parameter is "
+                "specified, as both would create categories that are all "
+                "zero.")
+
+    def _check_is_fitted(self):
+        if not self._fitted:
+            raise RuntimeError("Model must first be .fit()")
+
+    def _compute_drop_idx(self):
+        if self.drop is None:
+            return None
+        elif isinstance(self.drop, str) and self.drop == 'first':
+            return {feature: cp.array([0])
+                    for feature in self._encoders.keys()}
+        elif isinstance(self.drop, dict):
+            if len(self.drop.keys()) != len(self._encoders):
+                msg = ("`drop` should have as many columns as the number "
+                       "of features ({}), got {}")
+                raise ValueError(msg.format(len(self._encoders),
+                                            len(self.drop.keys())))
+            drop_idx = dict()
+            for feature in self.drop.keys():
+                cats = self._encoders[feature]
+                self.drop[feature] = Series(self.drop[feature])
+                if len(self.drop[feature]) != 1:
+                    msg = ("Trying to drop multiple values for feature {}, "
+                           "this is not supported.").format(feature)
+                    raise ValueError(msg)
+                if not self.drop[feature].isin(cats).all():
+                    msg = ("Some categories for feature {} were supposed "
+                           "to be dropped, but were not found in the encoder "
+                           "categories.".format(feature))
+                    raise ValueError(msg)
+                idx = cats.isin(self.drop[feature])
+                drop_idx[feature] = cp.asarray(cats[idx].index)
+            return drop_idx
+        else:
+            msg = ("Wrong input for parameter `drop`. Expected "
+                   "'first', None or a dict, got {}")
+            raise ValueError(msg.format(type(self.drop)))
+
+    def get_categories_(self):
+        """
+        Returns categories used for the one hot encoding in the correct order.
+
+        This copies the categories to the CPU and should only be used to check
+        the order of the categories.
+        """
+        return [self._encoders[f].to_array() for f in self._features]
+
+    def fit(self, X):
+        """
+        Fit OneHotEncoder to X.
+        Parameters
+        ----------
+        X : dask_cudf.DataFrame
+            The data to determine the categories of each feature.
+        Returns
+        -------
+        self
+        """
+        self._validate_keywords()
+        if type(self.categories) is str and self.categories == 'auto':
```

Comment:
One design pattern that I think might be a little easier to follow and maintain, and would also allow us to serialize multi-GPU models the same was as single GPU models, is to construct a single model from the features/encoders that you extract from the input. That way, we could eventually store that to disk and be able to load it back into a distributed model the same way. It also allows us to load estimators trained on a single GPU into a multi-GPU setting. 

You should be able to use the `DelayedPredictionMixin` and just store the singleGPU estimator as an instance variable.  That way, when transform and inverse_transform are invoked, you can just call `self._transform()` and `self.inverse_transform` to invoke the mixin.

Reply:
Good idea! I pushed a commit to address this. I could remove the bulk of `transform` and `inverse_transform` but I could not come up with a better way to re-use the single-GPU `fit` method like you did with `kmeans` model for instance. I ended up writing a custom `fit` for the multi-GPU model. If you can think of a better way to achieve this please tell me.

Reply:
I ended up refactoring slightly the single-node estimator to be able to reuse its fit function. I hope this solution is more maintainable than what I mentioned in the precedent comment.